duplicate DidOpenTextDocument

I'm using ra-multiplexer with neovim and lsp.
It all works fine when I open a new project for the first time. After closing and re-opening the editor (neovim) `ra-multiplexer` logs the following error:

```
[ERROR] [/my/project 1558707] [ERROR rust_analyzer::handlers::notification] duplicate DidOpenTextDocument: /my/project/src/main.rs
```

Everything seems to be working fine, but inlay_hints just disappear. Not sure if it's a ra-multiplexer bug or not.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 
+## [Unreleased]
+
+### Fixed
+- track open files by clients to avoid sending duplicate `textDocument/didOpen` and `textDocument/didClose` notifications
+
+
 ## [v0.2.3] - 2024-03-17
 
 ### Added

--- a/src/client.rs
+++ b/src/client.rs
@@ -9,7 +9,7 @@ use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
 use tokio::net::TcpStream;
 use tokio::sync::mpsc::error::SendError;
 use tokio::sync::{mpsc, Mutex};
-use tokio::{select, task};
+use tokio::task;
 use tracing::{debug, error, info, warn, Instrument};
 use uriparse::URI;
 
@@ -78,6 +78,7 @@ pub async fn process(
     }
 }
 
+#[derive(Clone)]
 pub struct Client {
     port: u16,
     sender: mpsc::Sender<Message>,
@@ -99,10 +100,6 @@ impl Client {
 
     pub fn port(&self) -> u16 {
         self.port
-    }
-
-    pub fn is_connected(&self) -> bool {
-        !self.sender.is_closed()
     }
 
     /// Send a message to the client channel
@@ -259,11 +256,10 @@ async fn connect(
     info!("initialized client");
 
     let (client, client_rx) = Client::new(port);
-    let (close_tx, close_rx) = mpsc::channel(1);
-    task::spawn(input_task(client_rx, close_rx, writer).in_current_span());
-    instance.add_client(client).await;
+    task::spawn(input_task(client_rx, writer).in_current_span());
+    instance.add_client(client.clone()).await;
 
-    task::spawn(output_task(reader, port, instance, close_tx).in_current_span());
+    task::spawn(output_task(reader, client, instance).in_current_span());
 
     Ok(())
 }
@@ -338,11 +334,7 @@ fn select_workspace_root<'a>(
 }
 
 /// Receive messages from channel and write them to the client input socket
-async fn input_task(
-    mut rx: mpsc::Receiver<Message>,
-    mut close_rx: mpsc::Receiver<Message>,
-    mut writer: LspWriter<OwnedWriteHalf>,
-) {
+async fn input_task(mut rx: mpsc::Receiver<Message>, mut writer: LspWriter<OwnedWriteHalf>) {
     // Unlike the output task, here we first wait on the channel which is going
     // to block until the language server sends a notification, however if
     // we're the last client and have just closed the server is unlikely to send
@@ -355,10 +347,7 @@ async fn input_task(
     // request was received but the client closed `close_rx` channel will be
     // dropped (unlike the normal rx channel which is shared) and the connection
     // will close without sending any response.
-    while let Some(message) = select! {
-        message = close_rx.recv() => message,
-        message = rx.recv() => message,
-    } {
+    while let Some(message) = rx.recv().await {
         if let Err(err) = writer.write_message(&message).await {
             match err.kind() {
                 // ignore benign errors, treat as socket close
@@ -376,9 +365,8 @@ async fn input_task(
 /// Read messages from client output socket and send them to the server channel
 async fn output_task(
     mut reader: LspReader<BufReader<OwnedReadHalf>>,
-    port: u16,
+    client: Client,
     instance: Arc<Instance>,
-    close_tx: mpsc::Sender<Message>,
 ) {
     loop {
         let message = match reader.read_message().await {
@@ -404,12 +392,12 @@ async fn output_task(
                 // <https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#shutdown>
                 let res = ResponseSuccess::null(req.id);
                 // Ignoring error because we would've closed the connection regardless
-                let _ = close_tx.send(res.into()).await;
+                let _ = client.send_message(res.into()).await;
                 break;
             }
 
             Message::Request(mut req) => {
-                req.id = req.id.tag(Tag::Port(port));
+                req.id = req.id.tag(Tag::Port(client.port));
                 if instance.send_message(req.into()).await.is_err() {
                     break;
                 }
@@ -435,7 +423,7 @@ async fn output_task(
             }
 
             Message::Notification(notif) if notif.method == "textDocument/didOpen" => {
-                if let Err(err) = instance.open_file(port, notif.params).await {
+                if let Err(err) = instance.open_file(client.port, notif.params).await {
                     warn!(?err, "error opening file");
                 }
             }
@@ -446,5 +434,9 @@ async fn output_task(
                 }
             }
         }
+    }
+
+    if let Err(err) = instance.cleanup_client(client).await {
+        warn!(?err, "error cleaning up after a client");
     }
 }

--- a/src/client.rs
+++ b/src/client.rs
@@ -1,4 +1,3 @@
-use std::collections::HashSet;
 use std::io::ErrorKind;
 use std::sync::Arc;
 
@@ -82,20 +81,12 @@ pub async fn process(
 pub struct Client {
     port: u16,
     sender: mpsc::Sender<Message>,
-    pub files: HashSet<String>,
 }
 
 impl Client {
     fn new(port: u16) -> (Client, mpsc::Receiver<Message>) {
         let (sender, receiver) = mpsc::channel(16);
-        (
-            Client {
-                port,
-                sender,
-                files: HashSet::default(),
-            },
-            receiver,
-        )
+        (Client { port, sender }, receiver)
     }
 
     pub fn port(&self) -> u16 {
@@ -105,13 +96,6 @@ impl Client {
     /// Send a message to the client channel
     pub async fn send_message(&self, message: Message) -> Result<(), SendError<Message>> {
         self.sender.send(message).await
-    }
-
-    pub fn get_status(&self) -> ext::Client {
-        ext::Client {
-            port: self.port,
-            files: self.files.iter().cloned().collect(),
-        }
     }
 }
 

--- a/src/client.rs
+++ b/src/client.rs
@@ -412,6 +412,12 @@ async fn output_task(
                 }
             }
 
+            Message::Notification(notif) if notif.method == "textDocument/didClose" => {
+                if let Err(err) = instance.close_file(client.port, notif.params).await {
+                    warn!(?err, "error closing file");
+                }
+            }
+
             Message::Notification(notif) => {
                 if instance.send_message(notif.into()).await.is_err() {
                     break;

--- a/src/client.rs
+++ b/src/client.rs
@@ -1,3 +1,4 @@
+use std::collections::HashSet;
 use std::io::ErrorKind;
 use std::sync::Arc;
 
@@ -80,12 +81,20 @@ pub async fn process(
 pub struct Client {
     port: u16,
     sender: mpsc::Sender<Message>,
+    pub files: HashSet<String>,
 }
 
 impl Client {
     fn new(port: u16) -> (Client, mpsc::Receiver<Message>) {
         let (sender, receiver) = mpsc::channel(16);
-        (Client { port, sender }, receiver)
+        (
+            Client {
+                port,
+                sender,
+                files: HashSet::default(),
+            },
+            receiver,
+        )
     }
 
     pub fn port(&self) -> u16 {
@@ -102,7 +111,10 @@ impl Client {
     }
 
     pub fn get_status(&self) -> ext::Client {
-        ext::Client { port: self.port }
+        ext::Client {
+            port: self.port,
+            files: self.files.iter().cloned().collect(),
+        }
     }
 }
 
@@ -420,6 +432,12 @@ async fn output_task(
 
             Message::ResponseError(res) => {
                 warn!(?res, "client responded with error");
+            }
+
+            Message::Notification(notif) if notif.method == "textDocument/didOpen" => {
+                if let Err(err) = instance.open_file(port, notif.params).await {
+                    warn!(?err, "error opening file");
+                }
             }
 
             Message::Notification(notif) => {

--- a/src/ext.rs
+++ b/src/ext.rs
@@ -100,6 +100,10 @@ pub async fn status(json: bool) -> Result<()> {
         for client in instance.clients {
             println!("    - Client");
             println!("      port: {}", client.port);
+            println!("      files:");
+            for file in client.files {
+                println!("        - {}", file);
+            }
         }
     }
     Ok(())

--- a/src/instance.rs
+++ b/src/instance.rs
@@ -1,4 +1,4 @@
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
 use std::io::ErrorKind;
 use std::ops::Deref;
 use std::path::Path;
@@ -43,11 +43,11 @@ pub struct Instance {
     /// Server's response to `initialize` request
     init_result: lsp::InitializeResult,
 
-    /// Sending messages to the language server instance
+    /// Handle for sending messages to the language server instance
     server: mpsc::Sender<Message>,
 
-    /// Send messages to clients
-    clients: Mutex<HashMap<u16, Client>>,
+    /// Data of associated clients
+    clients: Mutex<HashMap<u16, ClientData>>,
 
     /// Dynamic capabilities registered by the server
     dynamic_capabilities: Mutex<HashMap<String, lsp::Registration>>,
@@ -65,6 +65,33 @@ impl Drop for Instance {
     fn drop(&mut self) {
         // Make sure we're not leaking anything
         debug!("instance dropped");
+    }
+}
+
+/// Wrapper around client handle with additional data only the server instance
+/// knows about
+struct ClientData {
+    /// Handle for sending messages to clients
+    client: Client,
+
+    /// URIs of files currently opened by this client
+    files: HashSet<String>,
+}
+
+impl ClientData {
+    fn get_status(&self) -> ext::Client {
+        ext::Client {
+            port: self.client.port(),
+            files: self.files.iter().cloned().collect(),
+        }
+    }
+}
+
+impl Deref for ClientData {
+    type Target = Client;
+
+    fn deref(&self) -> &Self::Target {
+        &self.client
     }
 }
 
@@ -113,6 +140,10 @@ impl Instance {
             let _ = client.send_message(req.into()).await;
         }
 
+        let client = ClientData {
+            client,
+            files: HashSet::new(),
+        };
         if clients.insert(client.port(), client).is_some() {
             unreachable!("BUG: added two clients with the same port");
         }

--- a/src/instance.rs
+++ b/src/instance.rs
@@ -153,7 +153,9 @@ impl Instance {
     pub async fn cleanup_client(&self, client: Client) -> Result<()> {
         debug!("cleaning up client");
 
-        let Some(client) = self.clients.lock().await.remove(&client.port()) else {
+        let mut clients = self.clients.lock().await;
+
+        let Some(client) = clients.remove(&client.port()) else {
             // TODO This happens for example when the language server died while
             // client was still connected, and the client cleanup is attempted
             // with the instance being gone already. We should try notifying
@@ -161,11 +163,10 @@ impl Instance {
             bail!("client was not connected");
         };
 
-        for uri in client.files {
-            if let Err(err) = self.close_file(None, uri).await {
-                warn!(?err, "error closing file");
-            }
-        }
+        let files = client.files.into_iter().collect::<Vec<_>>();
+        self.close_all_files(&clients, files)
+            .await
+            .context("error closing files")?;
 
         Ok(())
     }
@@ -237,40 +238,53 @@ impl Instance {
         Ok(())
     }
 
-    /// Handle client closing a file with `textDocument/didOpen` or
-    /// disconnecting with file still opened
-    pub async fn close_file(&self, port: Option<u16>, uri: String) -> Result<()> {
-        let mut send_notification = true;
+    /// Handle `textDocument/didClose` client notification
+    pub async fn close_file(&self, port: u16, params: Value) -> Result<()> {
+        let params = serde_json::from_value::<lsp::DidCloseTextDocumentParams>(params)
+            .context("parsing params")?;
 
         let mut clients = self.clients.lock().await;
 
-        if let Some(port) = port {
-            clients
-                .get_mut(&port)
-                .context("no matching client")?
-                .files
-                .remove(&uri);
-        }
+        clients
+            .get_mut(&port)
+            .context("no matching client")?
+            .files
+            .remove(&params.text_document.uri);
 
-        for client in clients.values() {
-            if client.files.contains(&uri) {
-                debug!(?uri, "file still opened by another client");
-                send_notification = false;
-                break;
+        self.close_all_files(&clients, vec![params.text_document.uri])
+            .await
+    }
+
+    /// Handle closing many files at once and sending notifications for
+    /// definitely closed files
+    async fn close_all_files(
+        &self,
+        clients: &HashMap<u16, ClientData>,
+        files: Vec<String>,
+    ) -> Result<()> {
+        for uri in files {
+            let mut send_notification = true;
+
+            for client in clients.values() {
+                if client.files.contains(&uri) {
+                    debug!(?uri, "file still opened by another client");
+                    send_notification = false;
+                    break;
+                }
             }
-        }
 
-        if send_notification {
-            let params = lsp::DidCloseTextDocumentParams {
-                text_document: lsp::TextDocumentIdentifier { uri },
-            };
-            let notif = Notification {
-                jsonrpc: Version,
-                method: "textDocument/didClose".into(),
-                params: serde_json::to_value(params).unwrap(),
-            };
-            debug!(?notif, "last client closed file");
-            let _ = self.send_message(notif.into()).await;
+            if send_notification {
+                let params = lsp::DidCloseTextDocumentParams {
+                    text_document: lsp::TextDocumentIdentifier { uri },
+                };
+                let notif = Notification {
+                    jsonrpc: Version,
+                    method: "textDocument/didClose".into(),
+                    params: serde_json::to_value(params).unwrap(),
+                };
+                debug!(?notif, "last client closed file");
+                let _ = self.send_message(notif.into()).await;
+            }
         }
 
         Ok(())

--- a/src/instance.rs
+++ b/src/instance.rs
@@ -95,24 +95,42 @@ impl Instance {
         let mut clients = self.clients.lock().await;
         let dyn_capabilities = self.dynamic_capabilities.lock().await;
 
-        // Register all currently cached dynamic capabilities. We will drop the
-        // client response and we need to make sure the request ID is unique.
-        let id = RequestId::String("replay:registerCapabilities".into()).tag(Tag::Drop);
-        let params = lsp::RegistrationParams {
-            registrations: dyn_capabilities.values().cloned().collect(),
-        };
-        let req = Request {
-            id,
-            method: "client/registerCapability".into(),
-            params: serde_json::to_value(params).unwrap(),
-            jsonrpc: Version,
-        };
-        debug!(?req, "replaying server request");
-        let _ = client.send_message(req.into()).await;
+        if !dyn_capabilities.is_empty() {
+            // Register all currently cached dynamic capabilities if there are
+            // any. We will drop the client response and we need to make sure
+            // the request ID is unique.
+            let id = RequestId::String("replay:registerCapabilities".into()).tag(Tag::Drop);
+            let params = lsp::RegistrationParams {
+                registrations: dyn_capabilities.values().cloned().collect(),
+            };
+            let req = Request {
+                id,
+                method: "client/registerCapability".into(),
+                params: serde_json::to_value(params).unwrap(),
+                jsonrpc: Version,
+            };
+            debug!(?req, "replaying server request");
+            let _ = client.send_message(req.into()).await;
+        }
 
         if clients.insert(client.port(), client).is_some() {
             unreachable!("BUG: added two clients with the same port");
         }
+    }
+
+    /// Send cleanup messages and remove remove client for client map
+    pub async fn cleanup_client(&self, client: Client) -> Result<()> {
+        debug!("cleaning up client");
+
+        if self.clients.lock().await.remove(&client.port()).is_none() {
+            // TODO This happens for example when the language server died while
+            // client was still connected, and the client cleanup is attempted
+            // with the instance being gone already. We should try notifying
+            // these clients immediately and handling the cleanup separately.
+            bail!("client was not connected");
+        }
+
+        Ok(())
     }
 
     /// Send a message to the language server channel
@@ -255,22 +273,14 @@ async fn gc_task(
         interval.tick().await;
 
         for (key, instance) in &instance_map.lock().await.0 {
-            let mut message_readers = instance.clients.lock().await;
-
-            // Remove closed senders
-            //
-            // We have to check here because otherwise the senders only get
-            // removed when a message is sent to them which might leave them
-            // hanging forever if the language server is quiet and cause the
-            // GC to never trigger.
-            message_readers.retain(|_port, client| client.is_connected());
+            let clients = instance.clients.lock().await;
 
             let idle = instance.idle();
-            debug!(path = ?key.workspace_root, idle, readers = message_readers.len(), "check instance");
+            debug!(path = ?key.workspace_root, idle, clients = clients.len(), "check instance");
 
             if let Some(instance_timeout) = instance_timeout {
                 // Close timed out instance
-                if idle > i64::from(instance_timeout) && message_readers.is_empty() {
+                if idle > i64::from(instance_timeout) && clients.is_empty() {
                     info!(pid = instance.pid, path = ?key.workspace_root, idle, "instance timed out");
                     instance.close.notify_one();
                 }

--- a/src/instance.rs
+++ b/src/instance.rs
@@ -146,6 +146,41 @@ impl Instance {
         Ok(())
     }
 
+    /// Handle `textDocument/didOpen` client notification
+    pub async fn open_file(&self, port: u16, params: Value) -> Result<()> {
+        let params = serde_json::from_value::<lsp::DidOpenTextDocumentParams>(params)
+            .context("parsing params")?;
+        let uri = &params.text_document.uri;
+
+        let mut send_notification = true;
+
+        let mut clients = self.clients.lock().await;
+        for client in clients.values() {
+            if client.files.contains(uri) {
+                debug!(?uri, "file already open by another client");
+                send_notification = false;
+                break;
+            }
+        }
+
+        clients
+            .get_mut(&port)
+            .expect("no matching client")
+            .files
+            .insert(uri.clone());
+
+        if send_notification {
+            let notif = Notification {
+                jsonrpc: Version,
+                method: "textDocument/didOpen".into(),
+                params: serde_json::to_value(params).unwrap(),
+            };
+            let _ = self.send_message(notif.into()).await;
+        }
+
+        Ok(())
+    }
+
     pub fn get_status(&self) -> ext::Instance {
         let clients = self
             .clients

--- a/src/instance.rs
+++ b/src/instance.rs
@@ -122,12 +122,18 @@ impl Instance {
     pub async fn cleanup_client(&self, client: Client) -> Result<()> {
         debug!("cleaning up client");
 
-        if self.clients.lock().await.remove(&client.port()).is_none() {
+        let Some(client) = self.clients.lock().await.remove(&client.port()) else {
             // TODO This happens for example when the language server died while
             // client was still connected, and the client cleanup is attempted
             // with the instance being gone already. We should try notifying
             // these clients immediately and handling the cleanup separately.
             bail!("client was not connected");
+        };
+
+        for uri in client.files {
+            if let Err(err) = self.close_file(None, uri).await {
+                warn!(?err, "error closing file");
+            }
         }
 
         Ok(())
@@ -175,7 +181,7 @@ impl Instance {
         let mut clients = self.clients.lock().await;
         for client in clients.values() {
             if client.files.contains(uri) {
-                debug!(?uri, "file already open by another client");
+                debug!(?uri, "file is already opened by another client");
                 send_notification = false;
                 break;
             }
@@ -193,6 +199,46 @@ impl Instance {
                 method: "textDocument/didOpen".into(),
                 params: serde_json::to_value(params).unwrap(),
             };
+            debug!(?notif, "first client opened file");
+            let _ = self.send_message(notif.into()).await;
+        }
+
+        Ok(())
+    }
+
+    /// Handle client closing a file with `textDocument/didOpen` or
+    /// disconnecting with file still opened
+    pub async fn close_file(&self, port: Option<u16>, uri: String) -> Result<()> {
+        let mut send_notification = true;
+
+        let mut clients = self.clients.lock().await;
+
+        if let Some(port) = port {
+            clients
+                .get_mut(&port)
+                .context("no matching client")?
+                .files
+                .remove(&uri);
+        }
+
+        for client in clients.values() {
+            if client.files.contains(&uri) {
+                debug!(?uri, "file still opened by another client");
+                send_notification = false;
+                break;
+            }
+        }
+
+        if send_notification {
+            let params = lsp::DidCloseTextDocumentParams {
+                text_document: lsp::TextDocumentIdentifier { uri },
+            };
+            let notif = Notification {
+                jsonrpc: Version,
+                method: "textDocument/didClose".into(),
+                params: serde_json::to_value(params).unwrap(),
+            };
+            debug!(?notif, "last client closed file");
             let _ = self.send_message(notif.into()).await;
         }
 

--- a/src/lsp.rs
+++ b/src/lsp.rs
@@ -146,6 +146,7 @@ pub struct Unregistration {
     pub method: String,
 }
 
+/// Params for `textDocument/didOpen` notification
 #[derive(Serialize, Deserialize, Clone)]
 #[serde(rename_all = "camelCase")]
 pub struct DidOpenTextDocumentParams {
@@ -159,4 +160,17 @@ pub struct TextDocumentItem {
     pub language_id: String,
     pub version: u64,
     pub text: String,
+}
+
+/// Params for `textDocument/didClose` notification
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct DidCloseTextDocumentParams {
+    pub text_document: TextDocumentIdentifier,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct TextDocumentIdentifier {
+    pub uri: String,
 }

--- a/src/lsp.rs
+++ b/src/lsp.rs
@@ -145,3 +145,18 @@ pub struct Unregistration {
     pub id: String,
     pub method: String,
 }
+
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct DidOpenTextDocumentParams {
+    pub text_document: TextDocumentItem,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct TextDocumentItem {
+    pub uri: String,
+    pub language_id: String,
+    pub version: u64,
+    pub text: String,
+}

--- a/src/lsp/ext.rs
+++ b/src/lsp/ext.rs
@@ -165,6 +165,7 @@ pub struct Instance {
 #[serde(rename_all = "camelCase")]
 pub struct Client {
     pub port: u16,
+    pub files: Vec<String>,
 }
 
 #[derive(Serialize, Deserialize, Clone, Debug)]

